src/Simplex/Messaging/Agent/Store/SQLite/Schema.hs

```diff
@@ -69,35 +69,128 @@ sndQueues =
 connections :: Query
 connections =
   [sql|
-    CREATE TABLE IF NOT EXISTS connections(
+    CREATE TABLE connections(
       conn_alias TEXT NOT NULL,
       rcv_host TEXT,
       rcv_port TEXT,
       rcv_id BLOB,
       snd_host TEXT,
       snd_port TEXT,
       snd_id BLOB,
-      PRIMARY KEY(conn_alias),
-      FOREIGN KEY(rcv_host, rcv_port, rcv_id) REFERENCES rcv_queues(host, port, rcv_id),
-      FOREIGN KEY(snd_host, snd_port, snd_id) REFERENCES snd_queues(host, port, snd_id)
+      last_internal_msg_id INTEGER NOT NULL,
+      last_internal_rcv_msg_id INTEGER NOT NULL,
+      last_internal_snd_msg_id INTEGER NOT NULL,
+      PRIMARY KEY (conn_alias),
+      FOREIGN KEY (rcv_host, rcv_port, rcv_id) REFERENCES rcv_queues (host, port, rcv_id),
+      FOREIGN KEY (snd_host, snd_port, snd_id) REFERENCES snd_queues (host, port, snd_id)
     ) WITHOUT ROWID;
   |]
 
 messages :: Query
 messages =
   [sql|
-    CREATE TABLE IF NOT EXISTS messages(
-      agent_msg_id INTEGER NOT NULL,
+    CREATE TABLE messages(
       conn_alias TEXT NOT NULL,
-      timestamp TEXT NOT NULL,
-      message BLOB NOT NULL,
-      direction TEXT NOT NULL,
-      msg_status TEXT NOT NULL,
-      PRIMARY KEY(agent_msg_id, conn_alias),
-      FOREIGN KEY(conn_alias) REFERENCES connections(conn_alias)
+      internal_id INTEGER NOT NULL,
+      internal_ts TEXT NOT NULL,
+      internal_rcv_id INTEGER,
+      internal_snd_id INTEGER,
+      body BLOB NOT NULL,
+      PRIMARY KEY (conn_alias, internal_id),
+      FOREIGN KEY (conn_alias)
+        REFERENCES connections (conn_alias)
+        ON DELETE CASCADE,
+      FOREIGN KEY (conn_alias, internal_rcv_id)
+        REFERENCES rcv_messages (conn_alias, internal_rcv_id)
+        ON DELETE CASCADE
+        DEFERRABLE INITIALLY DEFERRED,
+      FOREIGN KEY (conn_alias, internal_snd_id)
+        REFERENCES snd_messages (conn_alias, internal_snd_id)
+        ON DELETE CASCADE
+        DEFERRABLE INITIALLY DEFERRED
+    ) WITHOUT ROWID;
+  |]
+
+-- table containing metadata of messages agent receives
+-- * external_snd_id
+--   id of the message at sender, i.e. 'external_snd_id' corresponds to 'internal_snd_id' from the sender's side.
+-- * rcv_status
+--   one of [semantically]: "received", "acknowledged to broker", "acknowledged to sender", changed in this order.
+-- * broker_ack_ts
```

Comment:
maybe ack_brocker_ts to be more different from broker_ts?

Reply:
good point